### PR TITLE
Non-experimental non-equi join conditions

### LINKED ISSUE #46254: Move unsupported conditions of INNER JOIN from ON to WHERE.
**Use case**

```
SELECT * FROM t1 INNER JOIN t2 ON t1.x = t2.y AND wtf(xyz)
```

should be translated to

```
SELECT * FROM t1 INNER JOIN t2 ON t1.x = t2.y WHERE wtf(xyz)
```

And

```
SELECT * FROM t1 INNER JOIN t2 ON wtf(xyz)
```

should be translated to

```
SELECT * FROM t1 CROSS JOIN t2 WHERE wtf(xyz)
```

**Describe the solution you'd like**

AST-based transformation.

### PR DESCRIPTION
**Pull request by vdimir**
### Changelog category (leave one):

- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
- Remove the `allow_experimental_join_condition` setting, allowing non-equi conditions by default

Close https://github.com/ClickHouse/ClickHouse/issues/46254

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Run only fuzzers related jobs (libFuzzer fuzzers, AST fuzzers, etc.)
- [ ]  Exclude: AST fuzzers
---
- [ ]  Do not test
- [ ]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Comment by robot-ch-test-poll1** [comment]
*This is an automated comment for commit 856fad2dd191a5fa5cbb778bb4a95dad7684ff85 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/69910/856fad2dd191a5fa5cbb778bb4a95dad7684ff85/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/b858595e54081fedae17bc5a6785bcb4a7794dc2/docs_check.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

**Comment by vdimir** [comment]
@novikd could you please take a look?

File: src/Planner/PlannerJoinTree.cpp [hunk 1]
```diff
@@ -1242,6 +1245,7 @@ void joinCastPlanColumnsToNullable(QueryPlan & plan_to_add_cast, PlannerContextP
 }
 
 JoinTreeQueryPlan buildQueryPlanForJoinNode(const QueryTreeNodePtr & join_table_expression,
+    const QueryTreeNodePtr & nearest_query_node,
```
Comment (novikd) [inline src/Planner/PlannerJoinTree.cpp:1248]: Do you actually need this parameter?

File: src/Planner/PlannerJoins.cpp [hunk 2]
```diff
@@ -358,21 +358,23 @@ void buildJoinClause(
...
+                const auto * node = appendExpression(joined_dag, join_expression, planner_context, join_node);
+                join_clause.addResidualCondition(node);
             }
             else
             {
                 throw Exception(
                     ErrorCodes::INVALID_JOIN_ON_EXPRESSION,
-                    "JOIN {} join expression contains column from left and right table, you may try experimental support of this feature by `SET allow_experimental_join_condition = 1`",
-                    join_node.formatASTForErrorMessage());
+                    "OUTER JOIN ON expression {} contains column from left and right table, which is not supported with join_use_nulls",
```
[the remark this answers is not in the document]
Comment (vdimir) [reply src/Planner/PlannerJoins.cpp]: I meant any of `LEFT/RIGHT/FULL`, changed it to print actual kind

File: src/Planner/PlannerJoinTree.cpp [hunk 3]
```diff
@@ -1635,6 +1687,28 @@ JoinTreeQueryPlan buildQueryPlanForJoinNode(const QueryTreeNodePtr & join_table_
         result_plan.unitePlans(std::move(join_step), {std::move(plans)});
     }
 
+    if (join_clauses_and_actions.mixed_join_expressions_actions)
```
[the remark this answers is not in the document]
Comment (vdimir) [reply src/Planner/PlannerJoinTree.cpp]: If we haven't moved expression into join, we can just add this expression as a filter after join, added comment

File: src/Planner/PlannerJoinTree.cpp (hunk 1 above)
Comment (vdimir) [reply src/Planner/PlannerJoinTree.cpp]: Indeed, initially I wanted not only to move expression from `ON` to `WHERE` here but in the opposite direction as well, so I needed to have corresponding query scope to access to `WHERE` query tree. However, let's avoid implementing in in that PR, maybe revisit after JoinStep refactoring, so here only focus on removal of `allow_experimental_join_condition`, so I've reverted that part of changes

File: src/Planner/PlannerJoinTree.cpp [hunk 4]
```diff
@@ -1274,23 +1278,53 @@ JoinTreeQueryPlan buildQueryPlanForJoinNode(const QueryTreeNodePtr & join_table_
...
         join_clauses_and_actions.right_join_expressions_actions.appendInputsForUnusedColumns(right_plan.getCurrentDataStream().header);
-        auto right_join_expressions_actions_step = std::make_unique<ExpressionStep>(right_plan.getCurrentDataStream(), std::move(join_clauses_and_actions.right_join_expressions_actions));
+        appendSetsFromActionsDAG(join_clauses_and_actions.right_join_expressions_actions, right_join_tree_query_plan.useful_sets);
+
+        QueryPlanStepPtr right_join_expressions_actions_step;
+        if (can_move_out_residuals && !right_pre_filters.empty())
+            right_join_expressions_actions_step = std::make_unique<FilterStep>(right_plan.getCurrentDataStream(), std::move(join_clauses_and_actions.right_join_expressions_actions), right_pre_filters[0]->result_name, false);
+        else
+            right_join_expressions_actions_step = std::make_unique<ExpressionStep>(right_plan.getCurrentDataStream(), std::move(join_clauses_and_actions.right_join_expressions_actions));
         right_join_expressions_actions_step->setStepDescription("JOIN actions");
```
[the remark this answers is not in the document]
Comment (vdimir) [reply src/Planner/PlannerJoinTree.cpp]: https://github.com/ClickHouse/ClickHouse/blob/d17d4c6d5c3edd6c2359df5f9807881ae85ae401/src/Planner/PlannerJoinTree.cpp#L1305-L1319

**Comment by vdimir** [comment]
I already merged resolved conflicts with #68682, so now I need either resolve conflicts again, or fix bug and resubmit, keeping this on hold, deciding what is easier...

File: src/Planner/PlannerJoinTree.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (vdimir) [reply src/Planner/PlannerJoinTree.cpp]: Tried to do it with helper function `forJoinSides` (it's quite common pattern to call some function in such way in joins), but seems it's a bit complicated, just extracted lambda and removed `forJoinSides`

**Pull request merged**